### PR TITLE
INTEXT-220: Add `WindowingOffsetManager`

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INTEXT-220

**Cherry-pick to 1.3.x**

### CODE REVIEW

File: src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+import reactor.rx.Stream;
+import reactor.rx.Streams;
+import reactor.rx.stream.GroupedStream;
+
+/**
+ * An {@link OffsetManager} that aggregates writes over a time or count window, using an underlying delegate to
+ * do the actual operations. Its purpose is to reduce the performance impact of writing operations
+ * wherever this is desirable.
+ * <p>
+ * Either a time window or a number of writes can be specified, but not both.
```
Comment (mbogoevici) [inline src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java]: This doesn't seem to be true anymore - the windowing operation in Reactor takes both count and time. The mutual exclusion was unique to Rx.

Comment (artembilan) [inline src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java]: Oh! I missed this JavaDocs.
Will fix after all other your feedbacks.

Thank you for reaching me!
And yes: good morning! 
:smile:

Comment (mbogoevici) [inline src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java]: I can fix on merge too :)

**Comment by mbogoevici** [comment]
Looks good to me. I would add an integration test showing that it works properly with the listener, updated offsets are persistent, etc - extending from `AbstractOffsetManagerTests`.

**Comment by artembilan** [comment]
> I would add an integration test

Yep! That's good catch.
Looks like all these jumps around to different projects destroy the mind.
Will do!

BTW the Windows report for the performance:
```
StopWatch 'OffsetManagers Performance': running time (millis) = 44868
-----------------------------------------
ms     %     Task name
-----------------------------------------
36469  081%  Just KafkaTopicOffsetManager
08165  018%  Just KafkaNativeOffsetManager
00128  000%  Window KafkaTopicOffsetManager
00106  000%  Window KafkaNativeOffsetManager
```

I can add it into the commit message as well for better informativity.

**Comment by mbogoevici** [comment]
Maybe we should document this elsewhere, publish some comparative results, etc. The commit message is too obscure, and I wouldn't focus on Windows performance solely, tbh.

**Comment by artembilan** [comment]
Pushed.
See the commit comment for more info.

File: src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+	@Override
+	public void close() throws IOException {
+		this.closed = true;
+		this.delegate.close();
+	}
+
+	@Override
+	public void flush() throws IOException {
+		if (this.offsets != null) {
+			this.offsets.awaitAndShutdown(this.shutdownTimeout, TimeUnit.MILLISECONDS);
```
Comment (mbogoevici) [inline src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java]: Haven't actually tried it, but doesn't this call prevent the `RingBufferProcessor` to be used afterwards? Essentially any call to `delete` or `reset` preventing any update afterwards?

Comment (artembilan) [inline src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java]: Correct, that's why I create a new one in the onComplete Consumer. See commit comments and other changes with it.

Comment (mbogoevici) [inline src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java]: Ok, thanks. I see it now.

**Comment by artembilan** [comment]
Thanks to @mbogoevici nit-picking we have observed the `Producer.close()` problem on the `this.ioThread.join();`. To be honest I don't see who kills that `deamon` in the `KafkaProducer`.

Excluding the `OffsetManager.destroy()` from the `stopWatch` measurement cardinally changes our performance report:
```
StopWatch 'OffsetManagers Performance': running time (millis) = 8648
-----------------------------------------
ms     %     Task name
-----------------------------------------
00558  006%  Just KafkaTopicOffsetManager
07858  091%  Just KafkaNativeOffsetManager
00135  002%  Window KafkaTopicOffsetManager
00097  001%  Window KafkaNativeOffsetManager
```
On Windows 7 x64 Intel i7-3770K. M-m-m. With usual SATA HDD.

The simple fix must be applied for the `OffsetManagerPerformanceTests` as a flip of two operators:
```
stopWatch.stop();
((DisposableBean) offsetManager).destroy();
```
@mbogoevici , promised in the private discussion to fix on merge.

**Comment by mbogoevici** [comment]
Squashed, rebased, merged w/https://github.com/spring-projects/spring-integration-kafka/commit/3de53a73abc98784172a472cb05ff31c4338b36b (master)

https://github.com/spring-projects/spring-integration-kafka/commit/3546b6d56574ffa9cf44b1fcd2f6a610e0c4daef (1.3.x)

**Pull request closed without merging**